python/semantic_kernel/connectors/ai/onnx/services/onnx_gen_ai_completion_base.py

```diff
@@ -0,0 +1,97 @@
+import os
+from collections.abc import AsyncGenerator
+from typing import Any
+
+import onnxruntime_genai as OnnxRuntimeGenAi
+
+from semantic_kernel.connectors.ai.onnx.onnx_gen_ai_prompt_execution_settings import OnnxGenAIPromptExecutionSettings
+from semantic_kernel.contents import ImageContent
+from semantic_kernel.exceptions import ServiceInitializationError, ServiceInvalidResponseError
+from semantic_kernel.kernel_pydantic import KernelBaseModel
+
+
+class OnnxGenAICompletionBase(KernelBaseModel):
+    """Base class for OnnxGenAI Completion services."""
+
+    model: Any
+    tokenizer: Any
+    tokenizer_stream: Any
+    enable_multi_modality: bool = False
+
+    def __init__(self, ai_model_path: str, **kwargs) -> None:
+        """Creates a new instance of the OnnxGenAICompletionBase class,loads model & tokenizer.
+
+        Args:
+            ai_model_path (str): Path to Onnx Model.
+            **kwargs: Additional keyword arguments.
+
+        Raises:
+            ServiceInitializationError: When model cannot be loaded
+
+        """
+        try:
+            json_gen_ai_config = os.path.join(ai_model_path + "/genai_config.json")
+            with open(json_gen_ai_config) as file:
+                config: dict = json.load(file)
+                if "vision" in config.get("model", {}):
+                    enable_multi_modality = True
+                    model = OnnxRuntimeGenAi.Model(ai_model_path)
+                    tokenizer = model.create_multimodal_processor()
+                    tokenizer_stream = tokenizer.create_stream()
+                else:
+                    enable_multi_modality = False
+                    model = OnnxRuntimeGenAi.Model(ai_model_path)
+                    tokenizer = OnnxRuntimeGenAi.Tokenizer(model)
+                    tokenizer_stream = tokenizer.create_stream()
+
+        except Exception as e:
+            raise ServiceInitializationError(f"Failed to initialize OnnxTextCompletion service: {e}")
+
+        super().__init__(
+            model=model,
+            tokenizer=tokenizer,
+            tokenizer_stream=tokenizer_stream,
+            enable_multi_modality=enable_multi_modality,
+            **kwargs,
+        )
+
+    async def _prepare_input_params(
+        self, prompt: str, settings: OnnxGenAIPromptExecutionSettings, image: ImageContent | None = None
+    ) -> Any:
+        params = OnnxRuntimeGenAi.GeneratorParams(self.model)
+        params.set_search_options(**settings.prepare_settings_dict())
+        if not self.enable_multi_modality:
+            input_tokens = self.tokenizer.encode(prompt)
+            params.input_ids = input_tokens
+        else:
+            if image is not None:
+                # With the use of Pybind there is currently no way to load images from bytes
+                # We can only open images from a file path currently
+                image = OnnxRuntimeGenAi.Images.open(str(image.uri))
+            input_tokens = self.tokenizer(prompt, images=image)
```

Comment:
Does it need to call the `encode` method, i.e. `self.tokenizer.encode(prompt, images=image)`?

Reply:
I thought i could unify the methods but after some testing and checking the examples in ONNX i wasn't successful. I think it has to do with the Multimodal Processor and the normal Tokenizer, that they are implemented in a different way and expect it like this.

1. https://github.com/microsoft/onnxruntime-genai/blob/main/examples/python/phi3v.py
2. https://github.com/microsoft/onnxruntime-genai/blob/main/examples/python/phi3-qa.py

Reply:
What I meant here is that `self.tokenizer` is an object. We probably should not call an object directly. Please verify.

Reply:
Honestly the there is no Documentation yet for the MultiProcessorClass in Onnx, but after checking the Source Code, i think there is no Option.

From what i can see here they binded the Process of Tokenizing an Image on the `__call__` method in python, so i think there is no way not calling the Object ? 

[Source](https://github.com/microsoft/onnxruntime-genai/blob/2f84f3d1f4dcd0c85bcdfd08442d97fd5aa8daa3/src/python/python.cpp#L494)

Would it make a difference if i would use `self.tokenizer.__call__(args)` ?

Reply:
Based on the pybind11 binding you shared, the MultiModalProcessor class has a `__call__` method, which allows you to call an instance of this class as if it were a function.